File: combinatorics/resource/ST/project_5/bai_toan_4/simple_graph/amap_to_am.cpp
#include <iostream>
#include <vector>
#include <map>
#include <sstream>

using namespace std;

// Ham chuyen ban do ke sang ma tran ke
vector<vector<int>> amapToMatrix(const map<int, map<int, int>>& amap, int n) {
    vector<vector<int>> matrix(n, vector<int>(n, 0)); // Ma tran n x n

    for (const auto& [i, neighbors] : amap) {
        for (const auto& [j, w] : neighbors) {
            matrix[i][j] = w;  // Gan trong so vao vi tri [i][j]
        }
    }

    return matrix;
}

int main() {
    int n;
    cout << "Nhap so dinh: ";
    cin >> n;
    cin.ignore();

    map<int, map<int, int>> amap;
    cout << "Nhap ban do ke (dinh: (dinh,trongso)...):\n";

    for (int i = 0; i < n; ++i) {
        string line;
        getline(cin, line);
        if (line.find(":") == string::npos) continue;

        stringstream ss(line);
        string u_str, rest;
        getline(ss, u_str, ':');
        int u = stoi(u_str);
        amap[u] = {};

        while (ss >> rest) {
            int v, w;
            sscanf(rest.c_str(), "(%d,%d)", &v, &w);
            amap[u][v] = w;
        }
    }

    auto matrix = amapToMatrix(amap, n);

    cout << "Ma tran ke:\n";
    for (const auto& row : matrix) {
        for (int val : row) {
            cout << val << " ";
        }
        cout << endl;
    }

    return 0;
}

File: combinatorics/resource/ST/project_5/bai_toan_4/simple_graph/amap_to_am.py
# Ham chuyen ban do ke sang ma tran ke
def amap_to_matrix(amap, n):
    matrix = [[0] * n for _ in range(n)]  # Khoi tao ma tran toan 0

    for i in amap:
        for j in amap[i]:
            matrix[i][j] = amap[i][j]  # Ghi trong so vao ma tran

    return matrix

# Nhap du lieu tu ban phim
n = int(input("Nhap so dinh: "))
amap = {}

print("Nhap ban do ke (dinh: (dinh,trongso)...):")
for _ in range(n):
    line = input()
    if ':' not in line:
        continue
    u_str, rest = line.split(':')
    u = int(u_str.strip())
    amap[u] = {}
    tokens = rest.strip().split()
    for token in tokens:
        v, w = map(int, token.strip("()").split(","))
        amap[u][v] = w

# Chuyen doi va in ma tran ke
matrix = amap_to_matrix(amap, n)

print("Ma tran ke:")
for row in matrix:
    print(" ".join(map(str, row)))